Zend/zend_gc.c

```diff
@@ -1162,6 +1162,10 @@ ZEND_API int zend_gc_collect_cycles(void)
 				zend_array *arr = (zend_array*)p;
 
 				GC_TYPE(arr) = IS_NULL;
+
+				/* GC may destroy arrays with rc>1. This is valid and safe. */
+				HT_ALLOW_COW_VIOLATION(arr);
+
```

Comment:
I think, it's better to reset refcounter, instead if setting magic flag.
#if ZEND_DEBUG
    /* GC may destroy arrays with rc>1. This is valid and safe. */
    GC_REFCOUNT(arr) = 1;
#endif